tests/lib/prepare-snapd.sh

```diff
@@ -0,0 +1,209 @@
+        echo "Failed to start ${SVC_NAME}" 1>&2
+        systemctl status "${SVC_NAME}" || true
+        return 1
+    fi
+
+    # Wait until ssh is ready
+    if ! wait_for_ssh "${SVC_NAME}"; then
+        echo "===== SERIAL PORT OUTPUT ======" 1>&2
+        cat "${work_dir}/serial.log" 1>&2
+        echo "===== END OF SERIAL PORT OUTPUT ======" 1>&2
+        return 1
+    fi
+
+    # Wait for the snap command to become ready
+    nested_wait_for_snap_command
+}
+
+install_core22_deps() {
+    sudo apt update -qq
+
+    # these should already be installed in GCE and LXD images with the google/lxd-nested 
+    # backend, but in qemu local images from qemu-nested, we might not have them
+    sudo apt install psmisc fdisk snapd mtools ovmf qemu-system-x86 sshpass whois -yqq
+
+    # TODO: https://bugs.launchpad.net/snapd/+bug/1712808
+    # There is a bug in snapd that prevents udev rules from reloading in privileged containers
+    # with the following error message: 'cannot reload udev rules: exit status 1' when installing
+    # snaps. However it seems that retrying the installation fixes it
+    if ! sudo snap install snapcraft --classic; then
+        echo "FIXME: snapcraft install failed, retrying"
+        sudo snap install snapcraft --classic
+    fi
+    sudo snap install ubuntu-image --classic
+}
+
+download_core22_snaps() {
+    local snap_branch="$1"
+
+    # get the model
+    curl -o ubuntu-core-amd64-dangerous.model https://raw.githubusercontent.com/snapcore/models/master/ubuntu-core-22-amd64-dangerous.model
+
+    # download neccessary images
+    snap download pc-kernel --channel=22/${snap_branch} --basename=upstream-pc-kernel
+    snap download pc --channel=22/${snap_branch} --basename=upstream-pc-gadget
+    snap download snapd --channel=${snap_branch} --basename=upstream-snapd
+}
+
+create_cloud_init_cdimage_config() {
+    local CONFIG_PATH=$1
+    cat << 'EOF' > "$CONFIG_PATH"
+#cloud-config
+datasource_list: [NoCloud]
+users:
+  - name: ubuntu
+    sudo: "ALL=(ALL) NOPASSWD:ALL"
+    lock_passwd: false
+    plain_text_passwd: 'ubuntu'
+  - name: external
+    sudo: "ALL=(ALL) NOPASSWD:ALL"
+    lock_passwd: false
+    plain_text_passwd: 'ubuntu'
+  - name: test
+    sudo: "ALL=(ALL) NOPASSWD:ALL"
+    lock_passwd: false
+    plain_text_passwd: 'ubuntu'
+    uid: "12345"
+
+bootcmd:
+  - mkdir -p /var/lib/console-conf
+  - touch /var/lib/console-conf/complete
+
+EOF
+}
+
+prepare_core22_cloudinit() {
+    # unpack gadget
+    gadgetdir=/tmp/gadget-workdir
+    unsquashfs -d $gadgetdir upstream-pc-gadget.snap
+
+    # add the cloud.conf to the gadget
+    create_cloud_init_cdimage_config "${gadgetdir}/cloud.conf"
+
+    # repack kernel snap
+    snap pack --filename=pc-gadget.snap "$gadgetdir"
+    rm -r $gadgetdir
+}
+
+build_core22_snap() {
+    local project_dir="$1"
+    local current_dir="$(pwd)"
+    
+    # run snapcraft
+    (
+        cd "$project_dir"
+        sudo snapcraft --destructive-mode
```

Comment:
How do we know if we are in the right Ubuntu release to build in destructive mode? I think we need some check maybe by using lsb_release to avoid shooting us in the foot. I am also wondering where is the meaning of `runs-on: ubuntu-latest` defined.

Reply:
To know what ubuntu-latest refers to one can check the docs - however ubuntu-latest always points to the latest LTS release. So at this moment 20.04 - but it will switch to ubuntu 22.04 once it's released. Is it only a specific ubuntu release we can use? I can just change the runner to require a specific version.

The only reason I'm building in destructive mode is just for the speed of it, we are already executing in a VM, so no reason to spin a VM up to build in an isolated context. Unless I'm misunderstanding something?

Reply:
The problem is that we should not use `--destructive-mode` if not on the same LTS as the base - we would be getting dependencies from the wrong place. I actually thought we were using 22.04 here and I was asking to check that was actually the case to avoid problems in the future if the version changed.

How is this done to build core-initrd in the tests? I assume we are using 22.04 in the main branch, is that not the case?

Reply:
Ah then I understand, in core-initrd and core-base (main tests) we actually spin up a google instance that runs ubuntu 22.04 and then run ```--destructive-mode``` there. So we use 22.04 as build base in those cases. This would be the only case not doing that, which is then wrong. 

It won't work without the --destructive-mode as the LXD base for 22.04 doesn't exist AFAIK?

Reply:
Ok, so unfortunately there is no easy way of doing this right now. I need to manually spin up a jammy instance, and then do ```--destructive-mode``` inside that, it seems like it's the way of doing this for core22 bases right now, as no LXD image exists for core22, only for jammy classic. I'll have to modify this to run in a jammy VM.